subsys/bluetooth/mesh/light_hsl_srv.c

```diff
@@ -0,0 +1,549 @@
+					      net_buf_simple_pull_le16(buf));
+}
+
+static void default_get_handle(struct bt_mesh_model *model,
+				     struct bt_mesh_msg_ctx *ctx,
+				     struct net_buf_simple *buf)
+{
+	if (buf->len != BT_MESH_LIGHT_HSL_MSG_LEN_GET) {
+		return;
+	}
+
+	default_rsp(model, ctx);
+}
+
+static void default_set_handle(struct bt_mesh_model *model,
+				     struct bt_mesh_msg_ctx *ctx,
+				     struct net_buf_simple *buf)
+{
+	if (buf->len != BT_MESH_LIGHT_HSL_MSG_LEN_DEFAULT) {
+		return;
+	}
+
+	default_set(model, ctx, buf);
+	default_rsp(model, ctx);
+}
+
+static void default_set_unack_handle(struct bt_mesh_model *model,
+				     struct bt_mesh_msg_ctx *ctx,
+				     struct net_buf_simple *buf)
+{
+	if (buf->len != BT_MESH_LIGHT_HSL_MSG_LEN_DEFAULT) {
+		return;
+	}
+
+	default_set(model, ctx, buf);
+}
+
+static void range_encode_status(struct net_buf_simple *buf,
+				struct bt_mesh_light_hsl_srv *srv,
+				enum bt_mesh_model_status status)
+{
+	bt_mesh_model_msg_init(buf, BT_MESH_LIGHT_HSL_OP_RANGE_STATUS);
+	net_buf_simple_add_u8(buf, status);
+	net_buf_simple_add_le16(buf, srv->hue.range.min);
+	net_buf_simple_add_le16(buf, srv->hue.range.max);
+	net_buf_simple_add_le16(buf, srv->sat.range.min);
+	net_buf_simple_add_le16(buf, srv->sat.range.max);
+}
+
+static void range_rsp(struct bt_mesh_model *model,
+			   struct bt_mesh_msg_ctx *rx_ctx,
+			   enum bt_mesh_model_status status)
+{
+	struct bt_mesh_light_hsl_srv *srv = model->user_data;
+
+	BT_MESH_MODEL_BUF_DEFINE(msg, BT_MESH_LIGHT_HSL_OP_RANGE_STATUS,
+				 BT_MESH_LIGHT_HSL_MSG_LEN_RANGE_STATUS);
+	range_encode_status(&msg, srv, status);
+	(void)bt_mesh_model_send(model, rx_ctx, &msg, NULL, NULL);
+}
+
+static enum bt_mesh_model_status range_set(struct bt_mesh_model *model,
+					   struct bt_mesh_msg_ctx *ctx,
+					   struct net_buf_simple *buf)
+{
+	struct bt_mesh_light_hsl_srv *srv = model->user_data;
+	struct bt_mesh_light_hsl_range hue, sat;
+
+	hue.min = net_buf_simple_pull_le16(buf);
+	hue.max = net_buf_simple_pull_le16(buf);
+	if (hue.max < hue.min) {
+		return BT_MESH_MODEL_ERROR_INVALID_RANGE_MIN;
+	}
+
+
+	sat.min = net_buf_simple_pull_le16(buf);
+	sat.max = net_buf_simple_pull_le16(buf);
+	if (sat.max < sat.min) {
+		return BT_MESH_MODEL_ERROR_INVALID_RANGE_MIN;
+	}
+
+	bt_mesh_light_hue_srv_range_set(&srv->hue, ctx, &hue);
+	bt_mesh_light_sat_srv_range_set(&srv->sat, ctx, &sat);
+
+	return BT_MESH_MODEL_SUCCESS;
+}
+
+static void range_get_handle(struct bt_mesh_model *model,
+				     struct bt_mesh_msg_ctx *ctx,
+				     struct net_buf_simple *buf)
+{
+	if (buf->len != BT_MESH_LIGHT_HSL_MSG_LEN_GET) {
+		return;
+	}
+
+	range_rsp(model, ctx, BT_MESH_MODEL_SUCCESS);
+}
+
+static void range_set_handle(struct bt_mesh_model *model,
+				     struct bt_mesh_msg_ctx *ctx,
+				     struct net_buf_simple *buf)
+{
+	if (buf->len != BT_MESH_LIGHT_HSL_MSG_LEN_RANGE_SET) {
+		return;
+	}
+
+	/* According to Test specification MMDL/SR/LHSLSE/BI-01-C, ignore
```

Comment:
I think the idea is that you should ignore illegal values, but if the application has some HW limitation that the client can't know about beforehand, it should use these return status codes. I haven't implemented any way for the user to set these limitations, so we end up never sending these error codes

Reply:
```
6.4.9.3.1
...
When a Light HSL Setup Server receives a Light HSL Range Set message or a Light HSL Range Set Unacknowledged message with values that cannot be accepted, it shall set the status of the operation to a value representing the reason why the values cannot be accepted, as defined in Table 7.2.
If the received message is a Light HSL Range Set message, the Light HSL Setup Server shall respond with a Light HSL Range Status message (see Section 6.4.6.2.9).
```
Looks like specification mandates sending status in any case.  I see test specification applies the opposite behavior for invalid params with ignoring them (4.4.4.22). The model spec and model test spec contradict each other.

Reply:
> The model spec and model test spec contradict each other.

I'm not sure I agree. There are two types of invalid values here: 
- Sending min > max is illegal from the spec, and should be treated as an invalid message and be ignored. 
- Sending min < application defined min value is not illegal, but since the app can't accept it, it should respond why it can't.

This implementation (and the other models with range) doesn't support the application defined min value concept, and will therefore never trigger the second case.